checkstyle.xml

```diff
@@ -208,5 +208,10 @@
         <module name="MatchXpath">
             <property name="query" value="//VARIABLE_DEF[./TYPE/IDENT[@text='var'] and not(./ASSIGN/EXPR/LITERAL_NEW)]"/>
         </module>
+
+        <!-- Prevent public records in API packages -->
+        <module name="MatchXpath">
+            <property name="query" value="//RECORD_DEF[./MODIFIERS/LITERAL_PUBLIC][//PACKAGE_DEF//DOT/IDENT[@text='api']]"/>
+        </module>
```

Comment:
You might want to add a custom error message here, as I don't think this will generate failures that are easy to understand if you don't know about the record rule. I don't know much about configuring checkstyle but custom messages seem possible: https://stackoverflow.com/a/18252466